### PR TITLE
Snow - Fix early return in Susong1999 model.

### PR DESCRIPTION
**Pull request by jomey**
Got taken away going through the available models and thought I learn about them by breaking them out and adding some tests.
Main bug found that the early return in Susong1999 did not match the required tuple form.

### CODE REVIEW

File: smrf/envphys/nasde_model/susong_1999.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+           - **percent_snow** (*numpy.array*) - Percent of the precipitation
+                that is snow in values 0.0-1.0.
+           - **rho_s** (*numpy.array*) - Snow density values in kg/m^3.
+        """
+        ps = np.zeros(precipitation.shape)
+        sd = np.zeros(ps.shape)
+
+        # if no precipitation return all zeros
+        if np.sum(precipitation) == 0:
+            return {'pcs': ps, 'rho_s': sd}
```
Comment (jomey) [inline smrf/envphys/nasde_model/susong_1999.py:97]: This was the bug, where two arrays instead of a tuple was returned.

**Review by scotthavens — APPROVED** [review]
This looks great Joe! Want to tackle `precipitation.py` next?

**Comment by scotthavens** [comment]
Looks like the `mock` module is not loaded in some circumstances in the tests for coverage and testing the source distributions.

File: smrf/tests/envphys/test_snow.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+class TestSnow(unittest.TestCase):
+    TEMPERATURES = np.random.uniform(low=-8, high=3, size=(12, 12))
+    PRECIPITATION = np.random.uniform(low=0, high=3, size=TEMPERATURES.shape)
+
+    def test_missing_model(self):
+        with self.assertRaises(ValueError):
+            Snow.phase_and_density(self.TEMPERATURES, self.PRECIPITATION, '')
+
+    @mock.patch.object(Susong1999, 'run')
+    def test_passes_arguments(self, model_patch):
```
[the remark this answers is not in the document]
Comment (jomey) [reply smrf/tests/envphys/test_snow.py:17]: Two things I am testing here:
1. That the mapped string from the config is also mapped to the correct snow model
2. The given two parameters are passed in the correct sequence

This is a perfect case for mocking the models in my perspective as unit test should be isolated as much as possible.

**Comment by jomey** [comment]
> Looks like the `mock` module is not loaded in some circumstances in the tests for coverage and testing the source distributions.

Do you have any preferences on how to include test gems for building wheels?

File: smrf/tests/envphys/test_snow.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (scotthavens) [reply smrf/tests/envphys/test_snow.py:17]: Sounds good

**Comment by scotthavens** [comment]
Could mock be placed as a test requirement in setup.py? The other option is to add the mock dependency to tox.ini and the build scripts in 'travis' directory.

**Comment by jomey** [comment]
The `tox.ini` is updated and `mock` is updated as a requirement there. I also have it in the `requirements_dev.txt` since my impression was this should be sufficient. So is `setup.py` a better place over the dev requirements? Mainly want to avoid adding it all over the place.

**Comment by jomey** [comment]
Other question I have is on why the tests are run during the building of the release? I can see an argument for platforms other than the travis tox runs, but running them again for the builds seems redundant. For the other platforms I would suggest that adding a pull request pipeline for those and then don't run the tests during the building at all.

**Comment by scotthavens** [comment]
> Other question I have is on why the tests are run during the building of the release? I can see an argument for platforms other than the travis tox runs, but running them again for the builds seems redundant. For the other platforms I would suggest adding a pull request pipeline for those and then don't run the tests during the building at all.

I was following 2 different lines of reasoning for this. First was the cibuildwheel recommendation was to run the test against the built package to ensure that it was built properly for that environment. This is what numpy and pandas do. Second, I have found that sometimes these catch different errors in the way the package is built that tox cannot mimic. This is essentially testing what we'd download from pypi which isn't the same as what tox tests (a git clone).

These tests on the build are only ran for PR's not for each branch so we're not running these all the time.

**Comment by jomey** [comment]
> I was following 2 different lines of reasoning for this. First was the cibuildwheel recommendation was to run the test against the built package to ensure that it was built properly for that environment. This is what numpy and pandas do. Second, I have found that sometimes these catch different errors in the way the package is built that tox cannot mimic. This is essentially testing what we'd download from pypi which isn't the same as what tox tests (a git clone).
> 
> These tests on the build are only ran for PR's not for each branch so we're not running these all the time.

Makes sense.

What do you think about using the `requirements_dev.txt` for the build scripts?
Basically replace this line:
```python3 -m pip install -r requirements.txt```
in all the scripts?

That way there would be less editing once a package is added.

**Comment by scotthavens** [comment]
That works, it'll just add a little bit of time to the build but would make it easer.

**Comment by jomey** [comment]
Ended up with a slightly different, but seems to work nicely. Yay! to the `extra_requires` setup.py section ....

**Pull request merged**